lib/Framework/BufferAccumulator.cpp

```diff
@@ -181,10 +181,13 @@ void Pothos::BufferAccumulator::require(const size_t numBytes)
     //dont do anything if the buffer is large enough
     if (queue.front().length >= numBytes) return;
 
+    //or if the accumulator is empty
+    if (queue.size() == 1 and _bytesAvailable == 0) return;
+
     //or the accumulator itself doesnt have enough bytes -- but can eventually have enough
     //we deduce this by checking if the requirement is larger than the actual buffer size
     //check the queue size because this optimization requires a single contiguous buffer
-    if (_bytesAvailable < numBytes and queue.size() == 1 and
+    if (_bytesAvailable < numBytes and queue.front().length == _bytesAvailable and
```

Comment:
Looks like this change caused some issues, there can be multiple buffers in queue that need to be squashed together by the logic proceeding this.

Reply:
Hmm. Without this change, it did lots of memcpy, which wasn't so efficient and didn't work with other memory managers (e.g., CUDA). I haven't looked at this in a while though. This is a pretty complicated area of the code though, so it's totally believable I didn't hit all the use cases! Do you have this working with PothosGPU now? My graphics card is too old so I haven't been able to try it out yet.

Reply:
It seemed like the lockup was as simple as connecting wave source -> wave monitor display. And for some reason the wave monitor triggering exacerbated it more than lets say the periodogram plotter. I didnt notice it either, until someone reported it when I updated windows installer. Ooops

I'm still trying to get to the bottom of it, but it was pretty clear reverting this change at least made it back to normal. Now I kept the circular buffer fixes, that should have been fine.

> Without this change, it did lots of memcpy, which wasn't so efficient and didn't work with other memory managers (e.g., CUDA)

So the lock up cases had several empty buffers queued up behind a buffer with like 8 bytes in it. And the call was looking for something like 16 bytes. So it just got stuck sitting on buffers.

Yea and I do want to make it better. Im pretty sure there is a happy middle ground here. There are undoubtedly some case where theres probably not any memcpy required, but the FIFO needs a little cleaning.

> Do you have this working with PothosGPU now? My graphics card is too old so I haven't been able to try it out yet.

@ncorgan has been churning on it https://github.com/pothosware/PothosGPU/commits/master

Reply:
I found a case that fails with my change applied but works without it ... involves fan-out to multiple different kinds of plotters.